src/sdk.js

```diff
@@ -95,24 +96,55 @@ class OpenForm {
     CSPNonce.setValue(CSPNonceValue);
     initialiseSentry(sentryDSN, sentryEnv);
 
-    // ensure that the basename has no trailing slash (for react router)
-    let pathname = basePath || window.location.pathname;
-    if (pathname.endsWith('/')) {
-      pathname = pathname.slice(0, pathname.length - 1);
+    let pathname;
+    if (!this.useHashRouting) {
+      pathname = basePath || window.location.pathname;
+      if (pathname.endsWith('/')) {
+        // ensure that the basename has no trailing slash (for react router)
+        pathname = pathname.slice(0, pathname.length - 1);
+      }
+    } else {
+      pathname = '';
     }
     this.basePath = pathname;
+    this.makeRedirect();
     this.calculateClientBaseUrl();
   }
 
+  makeRedirect() {
+    // Perform pre-redirect based on this action: this is decoupled from the backend
+    const query = new URLSearchParams(document.location.search);
+    const action = query.get('_of_action');
+    if (action) {
+      // as actionParams is in the form `q1%3Dquery1%26q2%3Dquery2`, we first URL decode and then parse it:
+      const actionParams = new URLSearchParams(decodeURIComponent(query.get('_of_action_params')));
+      query.delete('_of_action');
+      query.delete('_of_action_params');
+
+      const redirectPath = getRedirectPathname(action, actionParams);
+
+      let newUrl;
+      if (!this.useHashRouting) {
+        newUrl = `${window.location.origin}${window.location.pathname}/${redirectPath}${
+          query.size ? '?' + query : ''
+        }`;
```

Comment:
if `redirectPath` has query params *and* there's an uncontrolled query param like `foo=bar` from the CMS, you will end up with:

```
https://example.com/base-path/stap/step-1?submissionId=123?foo=bar
```
which is of course not the correct form. I think `getRedirectPathname` should return an object with the URL parts:

```
{
  "path": relativePath,
  "query": queryObj,
}
```
and then depending on the `useHashRouting` compose the final URL in this method.

Note that this (probably) works correctly with hash-based routing.

Reply:
> if `redirectPath` has query params _and_ there's an uncontrolled query param like `foo=bar` from the CMS, you will end up with:

But shouldn't we keep unrelated query params? If the CMS needs `foo` to work correctly and we strip it off, it can break things, can it?

Reply:
yes we need to keep it, but now the implementation is naive in that there will only be before-hash query params or after-hash query params while both can be present at the same time - the problem is the double `?` in the resulting URL if you *don't* use hash based routing.

So you need to:

1. Server loads URL with query params, possibly with CMS-required params
2. SDK initializes
3. SDK does `pick<query, '_of_action' | '_of_action_args'>`
4. SDK processes action & action args into the (relative) path & SDK-specific query params
5. SDK builds a new URL:
    - Join basePath + relative path
    - Merge (original) params with SDK-specific query params from 4.
    - Delete `_of_action` and `_of_action_args params
6. SDK does `pushState` with newly constructed URL
    - it retains the "unrelated" query params
    - it drops the private _action routing params
    - it merges in any query params required by the SDK (like `submission_uuid` and `?productId=...` for appointments)

The current implementation fails at 5.2 - the merging is not done properly

Reply:
I think the new implementation should cover this